File: cpp/#242_valid_anagram.cpp
// Given two strings s and t, return true if t is an anagram of s, 
// and false otherwise.

#include <iostream>
#include <unordered_map>

using namespace std;

class Solution
{
public:
    bool isAnagram(const string& s, const string& t)    {
        if (s.length() != t.length())
        {
            return false;
        }
        
        // Use a hashmap to count the frequency of characters in both strings
        unordered_map<char, int> freqMap;
        
        // Increment the frequency for characters in string s
        for (char c : s)
        {
            ++freqMap[c];
        }
        
        // Decrement the frequency for characters in string t
        for (char c : t)
        {
            // if (freqMap.find(c) == freqMap.end())
            // {
            //     return false; // Character not present in s
            // }

            --freqMap[c];

            // if (freqMap[c] < 0)
            // {
            //     return false; // More occurrences of character c in t than in s
            // }
        }
        
        // Check if all character frequencies are zero
        for (auto& entry : freqMap)
        {
            if (entry.second != 0)
            {
                return false;
            }
        }
        
        return true;
    }
};

int main()
{
    Solution sol;
    
    string s = "anagram";
    string t = "nagaram";
    
    if (sol.isAnagram(s, t))
    {
        cout << "t is an anagram of s." << endl;
    }
    else
    {
        cout << "t is ~not~ an anagram of s." << endl;
    }

    return 0;
}

File: python/#242_valid_anagram.py
from typing import List

class Solution:
    def isAnagram(self, s: str, t: str) -> bool:
        if len(s) != len(t): 
            return False
        
        #Creating a Dictionary
        hashMap = {}

        for c in s:
            hashMap[c] = hashMap.get(c, 0) + 1
            # if the key c is not already present in the dictionary, it sets the
            # default value to 0 before incrementing.

        for c in t:
            hashMap[c] = hashMap.get(c, 0) - 1 

        for key in hashMap:
            if hashMap[key] != 0:
                return False
        
        return True


def main():
    sol = Solution()

    s = "anagram"
    t = "nagaram"

    result1 = sol.isAnagram(s, t)
    print(f"Test Case 1: {result1}")  # Expected output: True


if __name__ == "__main__":
    main()